Token.py

```diff
@@ -21,12 +22,160 @@ def __init__(
         self.token_structure = token_structure
         self.category = category
 
+        # feature related fields
+        self.date_values = self.get_dates()
+        self.currency = self.get_currency()
+        self.address = self.get_address()
+        self.num_lable = self.get_num_label()
+        self.total_label = self.get_total_label()
+        self.date_label = self.get_date_label()
+        self.contains_digit = self.get_contains_digits()
+
     def __repr__(self):
         return self.text if self.text else str(self.token_structure)
 
     def __str__(self):
         return self.text if self.text else str(self.token_structure)
 
+    # returns true if has at least one digit, false otherwise
+    def get_contains_digits(self):
+        if self.text:
+            for c in self.text:
+                if c.isdigit():
+                    return True
+            return False
+
+    def get_date_label(self):
+        kw = ["date", "Date"]
+        if self.text:
+            for w in kw:
+                if w in self.text:
+                    return self.text
+
+    # tries to extract address from token
+    def get_address(self):
+        kw = [
+            "drive",
+            "dr",
+            "road",
+            "rd",
+            "lane",
+            "ln",
+            "ave",
+            "avenue",
+            "street",
+            "jalan",
+            "jln",
+            "boulevard",
+            "blvd",
+            "way",
+            "park",
+            "gate",
+            "crescent",
+            "cres",
+            "singapore",
+            "grove",
+            "grv"
+        ]
+        if self.text:
+            text_array = self.text.lower().split(" ")
+            for w in kw:
+                for t in text_array:
+                    if re.search("^" + w, t) and len(text_array) < 10:
+                        return self.text
+
+    # returns the text if "total" or some variant is contained in text and group is fewer than 5 words
+    def get_total_label(self):
+        kw = ["total", "Total"]
+        if self.text:
+            for w in kw:
+                if w in self.text and len(self.text.split(" ")) < 5:
+                    return self.text
+
+    # returns string for description of number, eg. account number, invoice number
+    def get_num_label(self):
+        kw = ["no.", "no:", "number", "num", "No.", "No:"]
+        if self.text:
+            text_array = self.text.split(" ")
+            for w in kw:
+                if w in text_array and text_array.index(w) > 0:
+                    return text_array[text_array.index(w) - 1]
```

Comment:
I think we could use the output here to create more detailed features in the feature engine. For example, there could be a boolean for whether a token aligns with a:
- Invoice number label
- Account number label
- PO number label


Currently if I understand correctly the feature engine only returns whether or not a token aligns with a token that has variations of the word number in it, which will classify the above tokens into the same class

Reply:
yes was thinking about this too but not sure if the invoices are consistent in their use of terms such as invoice and account and PO

Reply:
Yeah, we might need to create a list of terms that all reflect the same meaning (something like what you did for the different currencies)